Dockerfile

```diff
@@ -1,40 +1,27 @@
-FROM debian:jessie
-
-RUN apt-get update
-RUN apt-get install -y apt-transport-https
-RUN apt-get install -y ca-certificates
-RUN apt-get install -y bsdiff git curl
-
-ENV PACKAGE_NAME github.com/getlantern/autoupdate-server
-
-ENV WORKDIR /app
-RUN mkdir -p $WORKDIR
-
-ENV GO_VERSION go1.7.3
+FROM golang:1.12 as builder
 
-ENV GOROOT /usr/local/go
-ENV GOPATH /go
+WORKDIR $GOPATH/src/github.com/getlantern/autoupdate-server
+# Copying the mod and sum files, then running 'go mod download' creates a
+# separate layer of cached dependencies.
+COPY go.mod go.sum ./
+RUN GO111MODULE=on go mod download
+COPY . .
+RUN GO111MODULE=on go build -o /bin/autoupdate-server
+RUN GO111MODULE=on go build -o /bin/autoupdate-server-mock -tags mock
 
-RUN mkdir -p $GOPATH/bin $GOPATH/src $GOPATH/pkg
 
-ENV PATH $PATH:$GOROOT/bin:$GOPATH/bin
-
-ENV GO_PACKAGE_URL https://storage.googleapis.com/golang/$GO_VERSION.linux-amd64.tar.gz
-RUN curl -sSL $GO_PACKAGE_URL | tar -xvzf - -C /usr/local
-
-RUN curl https://glide.sh/get | bash
-
-ENV APPSRC_DIR /go/src/$PACKAGE_NAME
-ENV mkdir -p $APPSRC_DIR
-COPY ./ $APPSRC_DIR/
+# Running container
+FROM debian:jessie
 
-RUN cp $APPSRC_DIR/bin/entrypoint.sh /bin/entrypoint.sh
+RUN apt-get update && apt-get install -y \
+	ca-certificates
 
-RUN go build -o /bin/autoupdate-server $PACKAGE_NAME
-RUN go build -tags mock -o /bin/autoupdate-server-mock $PACKAGE_NAME
+COPY --from=builder /bin/autoupdate-server /bin/autoupdate-server
+COPY --from=builder /bin/autoupdate-server-mock /bin/autoupdate-server-mock
 
-VOLUME [ "/keys", $APPSRC_DIR, $WORKDIR ]
+COPY bin/entrypoint.sh /bin/entrypoint.sh
 
-WORKDIR $WORKDIR
+RUN mkdir /app
+VOLUME [ "/keys", "/app" ]
 
 ENTRYPOINT ["/bin/entrypoint.sh"]
```

Comment:
Summary of changes here:

I needed to upgrade to Go 1.12.  I used a [multistage build](https://docs.docker.com/develop/develop-images/multistage-build/) so that we can easily update our Go version without having to actually install it in the image.  I then removed all the stuff formerly used to install Go.

Reply:
The autoupdate server is running an older version of docker that doesn't support multistage builds, so to install this we'll need to upgrade docker on the autoupdate server. To be safe, I think it would be good to create a staging version of the autoupdate server to test that this new stuff can be deployed correctly.

The existing autoupdate server is running at 162.243.50.247 and you can ssh to it as `root`. Something that I don't love about it is that it's a little non-standard compared to our usual servers. It's not managed by salt, and it runs the app as a docker container. It has been reliable, so I don't see any prima facie reason to change its basic structure, but I do think it would be nice to have it managed by salt so that we can easily reproduce its configuration.

That's all a long-winded introduction to saying that I think it would be useful if you:

1. Create a salt configuration for configuring auto-update servers
2. Launch a new auto-update server
3. Deploy your branch to this auto-update server and make sure it works
4. Switch DNS to point update.getlantern.org at the new server
5. Retire the old auto-update server
6. Update support template to reference new auto-update server

Reply:
Nice!